docs: testing docs 🎉 (#7703)
artsy/eigen: docs/testing.md

- API Queries https://callstack.github.io/react-native-testing-library/docs/api-queries/

- We refer to this guide 👉 [How should I query?](https://callstack.github.io/react-native-testing-library/docs/how-should-i-query/) from `@testing-library/react-native` for querying components

Based on the [Guiding Principles](https://testing-library.com/docs/guiding-principles/), your test should resemble how users interact with your code (component, page, etc.) as much as possible.

Therefore we prefer using `screen.getByText`, `screen.getByDisplayValue`, etc, and as a last resort use a `testID` with `screen.getByTestId`.

In our code base there is some "Old fashioned" testing used for Class components for example since they can only be used using the old way in combination with QueryRenderer etc

The "new fashion" testing uses > relay v13 and we now prefer to use hooks whenever possible. We would for example use `useLazyLoadQuery` instead of a `QueryRenderer`.

- [`relay-test-utils`](https://relay.dev/docs/guides/testing-relay-components/) is the framework we use for testing relay components, creating mock environments etc.

> - We don’t use snapshot tests; they produce too much churn for too little value. It’s okay to test that a component doesn’t throw when rendered, but use [`extractText`](https://github.com/artsy/eigen/blob/4c7c9be69ab1c2095f4d2fed11a040b1bde6eba8/src/lib/tests/extractText.ts) (or similar) to test the actual component tree.

Reviewer: this can be removed

> - We don’t use snapshot tests; they produce too much churn for too little value. It’s okay to test that a component doesn’t throw when rendered, but use [`extractText`](https://github.com/artsy/eigen/blob/4c7c9be69ab1c2095f4d2fed11a040b1bde6eba8/src/lib/tests/extractText.ts) (or similar) to test the actual component tree.

Reviewer 2: ```suggestion
- We don’t use snapshot tests; they produce too much churn for too little value.
```

> Our preferred methods to use are marked with a ⭐️, while the ones we'd like to avoid are marked as ❗️ depracated.

Reviewer: ```suggestion
Our preferred methods to use are marked with a ⭐️, while the ones we'd like to avoid are marked as ❗️ deprecated.
```

## Test helpers and wrappers, when to use them, and good examples

> - `TestRendered` is a component we create, using the component we want to test. We use a TestRendered when for example we want to create a mock query. We don't always need to use a test renderer.

Reviewer: ```suggestion
- `TestRenderer` is a component we create, using the component we want to test. We use a TestRenderer when for example we want to create a mock query. We don't always need to use a test renderer.
```

New fashion: You can use useLazyLoadQuery inside a TestRendered which is the new relay hooks way for querying data and the way we generally want to swift towards.

New fashion Example: [ArtworkConsignments.tests.tsx](https://github.com/artsy/eigen/blob/6360ddb5304e6c5bb1dc207db13d5bf0a5d7d9b6/src/app/Scenes/Artwork/Components/ArtworkConsignments.tests.tsx#L28-L38)

> Old fashioned: You can use a TestRendered to wrap your component with a QueryRenderer and pass it the test query data.

Reviewer: ```suggestion
Old fashioned: You can use a TestRenderer to wrap your component with a QueryRenderer and pass the test query data.
```

Old fashion Example: [Inbox.tests.tsx](https://github.com/artsy/eigen/blob/c96dd0807555d69ca2e8655dc68085276d249080/src/app/Containers/Inbox.tests.tsx)

> - ⭐️ [`renderWithWrappers`](https://github.com/artsy/eigen/blob/main/src/app/tests/renderWithWrappers.tsx#L88-L103) Our default method for using a component. Can be used on all components. wraps the component with a number of wrappers such as TrackingProvider, GlobalStoreProvider, SafeAreaProvider, etc. Using `testing library`. See the list of wrappers [here](https://github.com/artsy/eigen/blob/main/src/app/tests/renderWithWrappers.tsx#L19-L25).

Reviewer: ```suggestion
- ⭐️ [`renderWithWrappers`](https://github.com/artsy/eigen/blob/main/src/app/tests/renderWithWrappers.tsx#L88-L103) Our default method for using a component. Can be used on all components. Wraps the component with a number of wrappers such as TrackingProvider, GlobalStoreProvider, SafeAreaProvider, etc. Using `testing library`. See the list of wrappers [here](https://github.com/artsy/eigen/blob/main/src/app/tests/renderWithWrappers.tsx#L19-L25).
```

- ⭐️ `renderWithHookWrappersTL` Wraps the component with a relay hook environment, and a Suspense fallback. Example: [Activity.tests.tsx](https://github.com/artsy/eigen/blob/1a611488042f6eccfc62862fddf7d06a17087f0e/src/app/Scenes/Activity/Activity.tests.tsx)

- ⭐️ `setupTestWrapperTL` ❗️ This function does not work with hooks ❗️ Uses react-native testing library.

For components that use relay requests.

Very similar to `renderWithWrappers`. The difference is that it will also resolve the first api request **only**.

If your component has more than one requests you should use `renderWithWrappers` instead, in combination with one of the functions that resolve api requests such as `resolveMostRecentRelayOperation` or `rejectMostRecentRelayOperation` Example: **\_**

This is an abstraction that is sometimes convenient.

Example of a component with one request [PartnerLocationSection.tests.tsx](https://github.com/artsy/eigen/blob/7703aa01103a06b69b650af11bc9903ab37b1c4b/src/app/Scenes/Partner/Components/PartnerLocationSection.tests.tsx#L21).

> `setupTestWrapperTL` can be used to replace the combination of **TestRendered** and **renderWithWrappers**.

Reviewer: ```suggestion
`setupTestWrapperTL` can be used to replace the combination of **TestRenderer** and **renderWithWrappers**.
```

It returns the function `renderWithRelay` which is similar to `renderWithWrappers`. Both `renderWithRelay` and `renderWithWrappers` use [render](https://testing-library.com/docs/angular-testing-library/api/#render) under the hood. I

> t gives you access to all the [queries](https://callstack.github.io/react-native-testing-library/docs/api-queries) such as getBy, getAllBy, queryBy, queryAllBy, findBy, findAllBy, etc.

Reviewer: ```suggestion
It gives you access to all the [queries](https://callstack.github.io/react-native-testing-library/docs/api-queries) such as getBy, getAllBy, queryBy, queryAllBy, findBy, findAllBy, etc.
```

❗️ It won't work if there is already a component using a query renderer. [eg?]

// TODO: rename setupTestWrapperTL to setupTestWrapper and setupTestWrapper to setupTestWrapperLegacy
// Decide if we want to use this pattern or not.

> Suggestion: Would it make sense to either remove this abstraction alltogether, or suggest it as the main course of action?

Reviewer: Not sure If these lines need to be kept here

Reviewer 2: lets keep them 🤷. we do that for code I guess :D

# Testing

So you want to write tests!

- We use `@testing-library/react-native` and our helper `renderWithWrappers`. In the past we used ReactTestRenderer which we try to now move away from.
- We refer to this guide 👉 [How should I query?](https://callstack.github.io/react-native-testing-library/docs/how-should-i-query/) from `@testing-library/react-native` for querying components

## Test helpers and wrappers, when to use them, and good examples

> - `extractTest` takes a component and returns any string that exists there, whether it's on a title, View or button. You can also check `extractTest.tests.tsx`.

Reviewer 3: (#minor) It might be worth ordering these functions by most important / most used to least

Author: That's what I tried to do, without being very strict about it, ordering them in order of how often do I see something being used / how important I think they are. But I'm happy to update the order if you have another suggestion 🙏 😊

> - `extractTest` takes a component and returns any string that exists there, whether it's on a title, View or button. You can also check `extractTest.tests.tsx`.

Reviewer: I think the function here and the file name has a typo, its `extractText` not `extractTest`.

Also I would suggest to remove this function (we use it in two places in eigen).

It works but it was used along with enzyme in order to extract the text from the elements. Now that we use `@testing-library/react-native` there is no need for that anymore since the library itself has the ability to query for text with [byText](https://callstack.github.io/react-native-testing-library/docs/api-queries#bytext)

Author: I've added this, thank you!

> - `renderWithWrappers` Can be used on all components. Our default method for using a component. wraps the component with a number of wrappers such as TrackingProvider, GlobalStoreProvider, SafeAreaProvider, etc. Using `testing library`.

Reviewer 4: **suggestion:** maybe we can link to the wrappers here https://github.com/artsy/eigen/blob/main/src/app/tests/renderWithWrappers.tsx#L18

Author: Done, thank you for the suggestion! 🙏

- `renderWithWrappersLEGACY`depracated - avoid this ❗️ uses ReactTestRenderer

> - `setupTestWrapperTL` For components that use relay requests. Uses react-native testing library. Renders a test component and resolves the most recent operation. An abstraction that sometimes is convenient. Example [PartnerLocationSection.tests.tsx](https://github.com/artsy/eigen/blob/7703aa01103a06b69b650af11bc9903ab37b1c4b/src/app/Scenes/Partner/Components/PartnerLocationSection.tests.tsx#L21). This function can be used instead of TestRendered and renderWithWrappers. Returns the function `renderWithRelay` which is similar to `renderWithWrappers` and gives you access to a number of properties such as `findWithTestID` etc. It won't work if there is already a component using a query renderer.

Reviewer 4: **issue:**

> Renders a test component and resolves the most recent operation.

I am not sure we do the second part here related to resolving the most recent operation. We do that separately I believe

Author: Thank you Mounir. Would you please rephrase this with a commit or a comment here in a way you think makes more sense?

Reviewer: setupTestWrapperTL is great if you don't want access to the relay environment for any request other than the first one. What it does is that it renders the component and resolves the **First** request.

Author: Added this, thank you

Reviewer 4: Sorry for the late reply, thanks for helping with the @gkartalis

> - `setupTestWrapperTL` For components that use relay requests. Uses react-native testing library. Renders a test component and resolves the most recent operation. An abstraction that sometimes is convenient. Example [PartnerLocationSection.tests.tsx](https://github.com/artsy/eigen/blob/7703aa01103a06b69b650af11bc9903ab37b1c4b/src/app/Scenes/Partner/Components/PartnerLocationSection.tests.tsx#L21). This function can be used instead of TestRendered and renderWithWrappers. Returns the function `renderWithRelay` which is similar to `renderWithWrappers` and gives you access to a number of properties such as `findWithTestID` etc. It won't work if there is already a component using a query renderer.

Reviewer: ```suggestion
- `setupTestWrapperTL` For components that use relay requests. Uses react-native testing library. Renders a test component and resolves the most recent operation. An abstraction that sometimes is convenient. Example [PartnerLocationSection.tests.tsx](https://github.com/artsy/eigen/blob/7703aa01103a06b69b650af11bc9903ab37b1c4b/src/app/Scenes/Partner/Components/PartnerLocationSection.tests.tsx#L21). This function can be used instead of TestRendered and renderWithWrappers. Returns the function `renderWithRelay` which is similar to `renderWithWrappers` (both use render from @testing-library/react-native under the hood) and gives you access to all the [queries](https://callstack.github.io/react-native-testing-library/docs/api-queries) coming from it. It won't work if there is already a component using a query renderer.
```

Reviewer: let me know if you don't agree with the suggested change

Author: I've added this, thank you!

  Better to explicitly use a QueryRenderer to make obvious what is there and minimize complexity.

- `setupTestWrapper` depracated - avoid this ❗️uses ReactTestRenderer . Renders a test component and resolves the most recent operation. An abstraction that sometimes is convenient.

// TODO: rename setupTestWrapperTL to setupTestWrapper and setupTestWrapper to setupTestWrapperLegacy
// Decide if we want to use this pattern or not.

> - `resolveMostRecentRelayOperation` resolves the query request. We always put it after rendering a component that has relay requests. Your rendered component makes a request and we use this function in tests to resolve it. Example file: [Inbox.tests.tsx](https://github.com/artsy/eigen/blob/c96dd0807555d69ca2e8655dc68085276d249080/src/app/Containers/Inbox.tests.tsx)

Reviewer: **suggestion(non-blocking):** up to you
```suggestion
- `resolveMostRecentRelayOperation` resolves the query request. We always use it after rendering a component that has relay requests. Your rendered component makes a request and we use this function in tests to resolve it. Example file: [Inbox.tests.tsx](https://github.com/artsy/eigen/blob/c96dd0807555d69ca2e8655dc68085276d249080/src/app/Containers/Inbox.tests.tsx)

> ℹ️ Note: There might be some cases that you would want to use `flushPromiseQueue` after that in order to wait for the request to be resolved. Use it only if after resolving the request returns the suspense fallback component and not the component. 
```

Reviewer 2: the note is kinda confusing, and i worry people might just try to add this without knowing why, or worse, when its not needed 😬.

Reviewer: happy to skip this

Author: I've added this under flushPromiseQueue docs, after waring the devs not to use it. Thank you!

> - `rejectMostRecentRelayOperation` for example if network is off / bad / you have a server error. Example: [FollowArtistLink.tests.tsx](**add link here**)

Reviewer 5: I see you linked to the examples, thanks for doing it!

- `flushPromiseQueue` This is a hack - try to avoid it if possible.

flushPromiseQueue will cancel all promise operations.

> This is usually called when promises have been used in a bad way, eg you return the promise instead of resolving it.

Reviewer 4: **issue:** I don't agree with the description of the `flushPromiseQueue` here. I think the method simply waits for a frame or 1/60 of a sec (depending on the js environment configuration where it's being called)

Author: Thank you Mounir. Would you please rephrase this with a commit or a comment here in a way you think makes more sense?

Author: @pvinis @gkartalis thoughts?

Reviewer 4: Sorry for the late reply. I would suggest something like so.

`flushPromiseQueue` is a helper we to wait for the UI for an animation frame. This can be useful in case you need to test an operation that happens asynchronously and that takes less than a frame

> flushPromiseQueue will cancel all promise operations.

Reviewer: **suggestion:** 
```suggestion
flushPromiseQueue will resolve all promise operations.
```

Author: done, thank you

Try to make a test run normally and if you are absolutely sure everything else is correct but the test still fails after a promise has been called (eg `resolveMostRecentRelayOperation` or `rejectMostRecentRelayOperation`) try adding this.

// TODO: remove flushPromiseQueue from everywhere [WIP]

> - `fetchMockResponseOnce` - do we really need this? What is it used for? 😇

Reviewer 4: **issue:** maybe we can remove this from here if it's not clear

Author: I would love it if anyone that has more context would add comments or make a commit to this PR so we can have more information about this function 🙏 This is why I left this here.

Reviewer: I think we can just add the reference to the third party [repo](https://github.com/jefflau/jest-fetch-mock#jest-fetch-mock) for this one since it is not exactly our helper.

Author: done, thank you

Reviewer 4: 🙏

- `mockTrackEvent` - mock data for tracking. We don't actually send this data to our data tracking provider for called analytics function `trackEvent`.

- `mockPostEventToProviders` expects a user behavior (eg button click) and mocks analytics for called analytics function `postEventToProviders`

- `mockFetchNotificationPermissions` mocking for the function `fetchNotificationPermissions`

- `mockTimezone` mocking for time/date eg with moment or luxon. Used for testing the date format.

- `mockNavigate` we test if navigation was called and with which parameters.

> - `waitUntil` Waits until something happens. RN Testing library has a similar component [waitFor](https://testing-library.com/docs/dom-testing-library/api-async/), that we could potentially replace this with.

Reviewer: I would remove this completely since it is a helper that is legacy

Reviewer 2: maybe then we should remove `renderWithWrappersLEGACY`? thought its used in a bunch of places still 🤔

Reviewer: + yep, agree

Author: I've added that we want to remove this, thank you!
